Sync interfaces to proposed changes

weierophinney opened the pull request:
This patch syncs the interfaces to those as proposed in php-fig/fig-standards#1022; the changes represent a BC break.

Changes include:

- Adoption of PHP 7 as the minimum supported version, in order to provide scalar type hints for arguments as required, and return type hints for all factory methods.
- `ResponseFactoryInterface::createResponse()` now defines an optional second argument, `string $reasonPhrase`.
- `ServerRequestFactoryInterface::createServerRequest()` now defines an optional third argument, `array $serverParams`.
- Removes `ServerRequestFactoryInterface::createServerRequestFromArray()`.
- `ResponseFactoryInterface` and `UploadedFileFactoryInterface` now extend `StreamFactoryInterface`.
- `RequestFactoryInterface` now extends `UriFactoryInterface` and `StreamFactoryInterface`.
- `ServerRequestFactoryInterface` now extends `UriFactoryInterface`, `StreamFactoryInterface`, and `UploadedFileFactoryInterface`.

The patch also updates all docblocks to match those in the proposed patch to the specification.

This patch SHOULD NOT be merged until the related patch on fig-standards is merged, to ensure it remains synchronized with any changes.

weierophinney committed:
Sync interfaces to proposed changes

This patch syncs the interfaces to those as proposed in
php-fig/fig-standards#1022; the changes represent a BC break.

Changes include:

- Adoption of PHP 7 as the minimum supported version, in order to
  provide scalar type hints for arguments as required, and return type
  hints for all factory methods.
- `ResponseFactoryInterface::createResponse()` now defines an optional
  second argument, `string $reasonPhrase`.
- `ServerRequestFactoryInterface::createServerRequest()` now defines an
  optional third argument, `array $serverParams`.
- Removes `ServerRequestFactoryInterface::createServerRequestFromArray()`.
- `ResponseFactoryInterface` and `UploadedFileFactoryInterface` now
  extend `StreamFactoryInterface`.
- `RequestFactoryInterface` now extends `UriFactoryInterface` and
  `StreamFactoryInterface`.
- `ServerRequestFactoryInterface` now extends `UriFactoryInterface`,
  `StreamFactoryInterface`, and `UploadedFileFactoryInterface`.

The patch also updates all docblocks to match those in the proposed
patch to the specification.

shadowhand commented on src/RequestFactoryInterface.php:
```diff
@@ -5,15 +5,19 @@
      * @return RequestInterface
      */
-    public function createRequest($method, $uri);
+    public function createRequest(string $method, $uri): RequestInterface;
```
I think we should make `$uri` strictly typed to `UriInterface`, because this is easy:

```php
$requestFactory->createRequest('GET', $requestFactory->createUri('http://example.com'));
```

now that `RequestFactoryInterface` extends `UriFactoryInterface`

mindplay-dk replied:
I'm actually a bit less sure about this one - compared with `UploadedFileFactoryInterface`, which I commented on regarding the same issue below, I think this one is different, because the URI factory has only one method. There's only way to create a URI, and that's from a string. We lose nothing by allowing `string` here and it's not competing with `UriInterface` for any functionality - the URI factory provides no additional options.

I think we can safely favor brevity in this case, and I think it makes sense to do so, since the request-factory is otherwise literally useless on it's own; using the request-factory would come with a hard dependency on the URI factory, which doesn't really make sense, when there's no advantage for the majority use-case. (marginal use-cases can of course still use the URI factory, if needed, but it's hard to think of a use-case.)

weierophinney replied:
I totally agree with @mindplay-dk here. Forcing the user to use the factory to also create the URI makes the request factory far less usable, for no good reason. Since the RequestFactory extends the UriFactory in this case, the `$uri` value can be passed on directly to the `createUri()` method in the case of a string.

shadowhand commented on src/ServerRequestFactoryInterface.php:
```diff
@@ -5,27 +5,22 @@
-     *  If no valid method or URI can be determined.
-     */
-    public function createServerRequestFromArray(array $server);
+    public function createServerRequest(string $method, $uri, array $serverParams = []): ServerRequestInterface;
```
Same here, strict `UriInterface`?

mindplay-dk replied:
Disagree on the same grounds as stated for request-factory.

shadowhand commented on src/UploadedFileFactoryInterface.php:
```diff
@@ -18,22 +18,22 @@
-     * @param string $clientFilename
-     * @param string $clientMediaType
+     * @param string|resource|StreamInterface $file Underlying file, PHP stream
+     *     resource, or StreamInterface representing the uploaded file content.
```
Maybe strictly `StreamInterface`? The ambiguity of `string $filename` vs `string $filecontent` worries me. We can remove that if we ensure that streams are always used, since stream factory has separate methods for each construction.

mindplay-dk replied:
Strongly agree with this - not even so much because of the ambiguity, but because of feature envy. This competes with the stream-factory interface, and directly leads to inconsistencies in consumer code.

weierophinney replied:
This makes sense to me. That said: I propose that `UploadedFileFactoryInterface` extend `StreamFactoryInterface`. The rationale is that you will need a stream instance to create an `UploadedFile` instance, which is the same rationale I used for extension in the other interface cases. Thoughts?

shadowhand replied:
Makes sense to me:

```php
$stream = $uploadFactory->createStream($content);
$upload = $uploadFactory->createUploadedFile($stream, ...);
```

weierophinney replied:
Ironically, I had the php-fig PR _already_ noting that `UploadedFileFactoryInterface` extends `StreamFactoryInterface`; I just hadn't taken the extra step of changing the intial argument to `createUploadedFile()` to a stream.

Updated in both patches now.

weierophinney committed:
Changes UploadedFileFactoryInterface to extend StreamFactoryInterface

Updates UploadedFileFactoryInterface to extend StreamFactoryInterface,
and to change the initial argument from:

```php
string|resource|StreamInterface $file
```

to:

```php
StreamInterface $stream
```

This removes ambiguity on how the stream is created; the user is now
responsible for using the stream factory (which the
`UploadedFileFactoryInterface` now extends, making it immediately
present) in order to generate the stream to compose as part of the
uploaded file.
